Add step-by-step documentation for package contributions

timholy opened the pull request:
All package developers want contributions from the community. However, learning git can be a huge hurdle (it was for me), and I think it's detrimental to ask users to first spend two hours reading so they can make a 3-line change to a package.

To make sure we don't scare away potential contributors, let's try to make their first few commits as painless as possible. Here's my attempt at making the process clearer. I'm well aware that this repeats material available elsewhere, but the point is to gather it into one place. I link out to the Git book for the actual "commit" step, but otherwise my intent was to make this self-contained (while adding links for more information).

The downside, of course, is that we have to keep this information up to date.

timholy committed:
Add step-by-step documentation for package contributions

nalimilan commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,147 @@ Once you do this, the package manager knows your GitHub user name and can config
+
+If you want to improve the online documentation of a package, the
+easiest approach (at least for small changes) is to use GitHub's
+online editting functionality. First, navigate to the respository's
```
editting -> editing

nalimilan commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,147 @@ Once you do this, the package manager knows your GitHub user name and can config
+
+- Create a branch for your changes: from the command line (not within
+  Julia), navigate to the package folder (e.g.,
+  ``~/.julia/v0.4/Foo/``) and type ``git checkout -b <newbranch>``,
```
I'd say "and create a new branch using ``git checkout -b <newbranch>``".

nalimilan commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,147 @@ Once you do this, the package manager knows your GitHub user name and can config
+  works as intended.
+
+- Run the package's tests and make sure they pass.  Usually you can do
+  this by running a file probably called ``runtests.jl`` from inside
```
"running a file" may not be completely obvious: people need to call `julia test/runtests.jl`.

nalimilan commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,147 @@ Once you do this, the package manager knows your GitHub user name and can config
+  + From the command line, type ``git push``.  This will add your new
+    commit(s) to the same pull request; you should see them appear
+    automatically on the page holding the discussion of your pull
+    request.
```
Maybe one word about how to squash when everything is ready? Maintainers will often ask for this.

timholy replied:
It's going to take more than one word...

nalimilan replied:
In the ideal situation where everything is clean, `git rebase -i master && git push --force` should be enough. For other cases, well... http://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

timholy replied:
I mean to walk people through it in a safe way. See commit that's about to appear.

nalimilan replied:
Great. LGTM.

nalimilan commented:
Very good idea!

timholy committed:
Add info re squashing, amplify running of tests [ci skip]

Also fixes typos, provides extra safety in the "git reset"
description, and addresses other remarks by @nalimilan.

timholy commented:
Thanks for a very helpful review, @nalimilan!

tkelman commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,223 @@ Once you do this, the package manager knows your GitHub user name and can config
+into the project's main repository. Because the online repository
+can't see the code on your private machine, you first *push* your
+changes to a publicly-visible location, your own online *fork* of
+package (hosted on your own personal GitHub account).
```
missing an article - fork of a/the package

tkelman commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,223 @@ Once you do this, the package manager knows your GitHub user name and can config
+
+- Create a branch for your changes: from the command line (not within
+  Julia), navigate to the package folder (e.g.,
+  ``~/.julia/v0.4/Foo/``) and create a new branch using ``git
```
I'd make a note to say "replace `~/.julia` with the result of `Pkg.dir()`", particularly for Windows users

tkelman commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,223 @@ Once you do this, the package manager knows your GitHub user name and can config
+
+In the description below, anything starting with ``Pkg.`` is meant to
+be typed at the Julia prompt; anything starting with ``git`` is meant
+to be typed at the command line.
```
On Windows this will have to be either via shell mode in Julia, or by explicitly running `git-bash.bat` or `git-cmd.bat` from the `Git` folder that currently comes with Julia installs. When we get rid of the bundled git we should make a note that command-line git should be installed separately for this kind of thing.

timholy replied:
I like the idea of using the shell mode; I should do that more often myself.

nalimilan replied:
Yeah, good idea for windows users. Maybe there should be a `Pkg.shell()` command that would start the shell mode in the package directory? That way, newcomers don't need to care about the shell nor about where the package is stored.

tkelman commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,223 @@ Once you do this, the package manager knows your GitHub user name and can config
+- Commit your changes: see `<http://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository>`_.
+
+- Submit your changes: from within julia, type
+  ``Pkg.submit("Foo")``. This will push your changes to your GitHub
```
This will require setting up an SSH key at this point, won't it?

timholy replied:
That's described in the section immediately above this one. But I can insert a reference.

tkelman replied:
oh right, yeah. a reminder for the forgetful would be good :)

tkelman commented on doc/manual/packages.rst:
```diff
@@ -346,6 +346,223 @@ Once you do this, the package manager knows your GitHub user name and can config
+
+- Create a new branch. This branch will hold your changes.
+- Make sure everything is committed to this branch.
+- ``git checkout master``. If this fails, *do not* procede further
```
proceed

timholy committed:
Switch command-line -> shell mode; fix typos

timholy committed:
Group "new package" steps under a global header

timholy commented:
Thanks, @tkelman! (Arg, forgot `[ci skip]` again...)

jiahao commented:
:+1:

I made you a figure. SVG in [this gist](https://gist.github.com/jiahao/93ec9b68be8cafec1caa) (Why does GitHub make working with SVG so difficult?!).

![githubworkflow](https://cloud.githubusercontent.com/assets/1732/6433661/f2b9aab0-c042-11e4-817d-c3b7edae14f2.png)

timholy commented:
Wow. Perhaps I should merge this and then you can integrate the figure(s). (I'm not exactly sure which steps are illustrated, but I am certain it would be very clear in context.)

jiahao commented:
Hmm, if it doesn't make sense to you then that's not a good sign...

tkelman commented:
This (and 3c7136e3697865cc1381b17199cb1a92a519f437) should be backport-worthy.

Luthaf commented:
That is great ! I though I'll have to write it myself. Thank you !

tkelman commented:
squashed and backported in 32d6c9a92c345971db6de5440c58d3c58c6d79f1

yakir12 commented:
THANK YOU!
